runtime/near-evm-runner/src/utils.rs

```diff
@@ -183,66 +187,497 @@ pub fn format_log(topics: Vec<H256>, data: &[u8]) -> std::result::Result<Vec<u8>
+                    let length = current_array_length.take();
+                    state = State::Close;
+                    token = Some(Type::Array {
+                        inner: Box::new(token.expect("if statement checks for some; qed")),
+                        length,
+                    });
+                    lexer.advance();
+                    array_depth += 1;
+                    continue;
+                } else {
+                    return Err(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName));
+                }
+            }
+            Token::BracketClose if array_depth == 10 => {
+                return Err(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName));
+            }
+            _ => {
+                return Err(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName));
+            }
+        };
+
+        token = Some(type_);
+        lexer.advance();
+    }
+
+    Ok(token.ok_or(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName))?)
+}
+
 pub fn near_erc721_domain(chain_id: U256) -> RawU256 {
     let mut bytes = Vec::with_capacity(70);
     bytes.extend_from_slice(
         &keccak("EIP712Domain(string name,string version,uint256 chainId)".as_bytes()).as_bytes(),
     );
-    bytes.extend_from_slice(b"NEAR");
-    bytes.extend_from_slice(&[0x01]);
+    let near = b"NEAR";
+    let near: RawU256 = keccak(&near).into();
+    bytes.extend_from_slice(&near);
+    let version = b"1";
+    let version: RawU256 = keccak(&version).into();
+    bytes.extend_from_slice(&version);
     bytes.extend_from_slice(&u256_to_arr(&chain_id));
     keccak(&bytes).into()
 }
 
-pub fn method_name_to_rlp(method_name: String) -> [u8; 4] {
+pub fn method_sig_to_abi(method_sig: &str) -> [u8; 4] {
     let mut result = [0u8; 4];
-    result.copy_from_slice(&keccak(method_name)[..4]);
+    result.copy_from_slice(&keccak(method_sig)[..4]);
     result
 }
 
+pub fn encode_address(addr: Address) -> Vec<u8> {
+    let mut bytes = vec![0u8; 12];
+    bytes.extend_from_slice(&addr.0);
+    bytes
+}
+
+pub fn encode_string(s: &str) -> Vec<u8> {
+    let mut bytes = vec![];
+    bytes.extend_from_slice(&keccak(s.as_bytes()).as_bytes());
+    bytes
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum Value {
+    Bytes(Vec<u8>),
+    List(Vec<Value>),
+}
+
+impl Decodable for Value {
+    fn decode(rlp: &Rlp<'_>) -> std::result::Result<Self, DecoderError> {
+        if rlp.is_list() {
+            Ok(Value::List(rlp.as_list()?))
+        } else {
+            Ok(Value::Bytes(rlp.decoder().decode_value(|bytes| Ok(bytes.to_vec()))?))
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct Arg {
+    #[allow(dead_code)]
+    pub name: String,
+    pub type_raw: String,
+    pub t: Type,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct Method {
+    pub name: String,
+    pub raw: String,
+    pub args: Vec<Arg>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct MethodAndTypes {
```

Comment:
same. It would make sense to put it all into a mod, if it has some context. E.g.:
```
mod MetaParsing {
pub struct MethodAndTypes
...
}